ledger/acctonline.go

```diff
@@ -621,104 +621,125 @@ func (ao *onlineAccounts) lookupOnlineAccountData(rnd basics.Round, addr basics.
+			return macct.GetOnlineAccountData(rewardsProto, rewardsLevel), nil
+		}
 
-	if persistedData.rowid != 0 { // if we read actual data return it
+		ao.accountsMu.RUnlock()
+		needUnlock = false
+
+		// No updates of this account in the in-memory deltas; use on-disk DB.
+		// As an optimization, we avoid creating
+		// a separate transaction here, and directly use a prepared SQL query
+		// against the database.
+		persistedData, err = ao.accountsq.lookupOnline(addr, rnd)
+		if err != nil || persistedData.rowid == 0 {
+			// no such online account, return empty
+			return ledgercore.OnlineAccountData{}, err
+		}
 		// lookupOnlineHistory does not fetch the account db round because of the following observation:
 		// 1. ao.onlineAccountsCache update happens with ao.accountsMu taken below and in postCommit
 		// 2. If we started reading the history (lookupOnlineHistory)
 		//   * before commitRound or while it is running => OK, read what is in DB and then add new entries in postCommit
 		//     - if commitRound deletes some history after, the cache has additional entries and updRound comparison gets a right value
 		//   * after commitRound but before postCommit => OK, read full history, ignore the update from postCommit in writeFront's updRound comparison
 		//   * after postCommit => OK, postCommit does not add new entry with writeFrontIfExist, but here all the full history is loaded
-		persistedDataHistory, err := ao.accountsq.lookupOnlineHistory(addr)
-		if err != nil {
+		persistedDataHistory, validThrough, err := ao.accountsq.lookupOnlineHistory(addr)
+		if err != nil || len(persistedDataHistory) == 0 {
 			return ledgercore.OnlineAccountData{}, err
 		}
-		err = func() error {
-			ao.accountsMu.Lock()
-			defer ao.accountsMu.Unlock()
-			ao.onlineAccountsCache.clear(addr)
-			if !ao.onlineAccountsCache.full() {
-				for _, data := range persistedDataHistory {
-					written := ao.onlineAccountsCache.writeFront(
-						data.addr,
-						cachedOnlineAccount{
-							baseOnlineAccountData: data.accountData,
-							updRound:              data.updRound,
-						})
-					if !written {
-						err = fmt.Errorf("failed to write history of acct %s for round %d into online accounts cache", data.addr.String(), data.updRound)
-						return err
+		// 3. If we fishied reading the history (lookupOnlineHistory)
+		//   * db has not advanced => OK
+		//   * after commitRound but before postCommit => OK, the cache gets populated and postCommit updates the new entry
+		//   * after commitRound and after postCommit => problem, postCommit does not add a new entry but the cache gets constructed by misses the latest entry
+		// In order to resolve this lookupOnlineHistory returns dbRound value and determine what happened
+		// We cannot handle case 3.2 without a huge mess with lock (can compare currentDeltaLen == len(ao.deltas) as an indication postCommit not called yet)
+		// So handle 3.1 here and retry for 3.2 and 3.3
+		if validThrough == currentDbRound {
```

Comment:
This check doesn't fix anything. We need to read `ao.cachedDbRound` under the mutex lock below.

Reply:
no, we need to check if db have not advanced while we are in lookup function. The only thing it needs to be under a lock to ensure postCommit does not update the cache while we are here.

Reply:
Fixed. The delay size does not matter unless we know validThrough is the same as the dbRound, or commit happened but `postCommit` not entered. Having a lock for this comparison prevents `postCommit` to enter, and the content of the history might be missing one element (see explanation in the comment), and postCommit will put this missed data into the cache when it gets a chance for execution.